### PR TITLE
Display search results when searching from a `MacosSearchField` (closes #216)

### LINKED ISSUE #216: Show a dropdown area with search results in `MacosSearchField`

### PR DESCRIPTION
**Pull request by whiplashoo**
## Adds a proper search functionality to the `MacosSearchField` widget.

| Dark Theme                                 | Light Theme                                |
| ------------------------------------------ | ------------------------------------------ |
|  |  |

A user can provide a `results` property of `SearchResultItems` to the search field constructor, which can be as simple as a list of strings to search from. Typing in the field opens a dropdown below (or above) it with the search results shown in macOS-style. Then, the `onResultSelected` callback can be used to retrieve the user-selected result. 

### Usage:

```dart
MacosSearchField(
  results: countries.map((e) => SearchResultItem(e)).toList(),
  placeholder: 'Search for a country...',
  onResultSelected: (resultItem) {
    debugPrint(resultItem.searchKey);
  },
),
```

### Screencast:

https://user-images.githubusercontent.com/9117427/167090368-6678eb2d-f3b5-4937-8c7b-36e96d96b0bb.mp4

Selecting "Belize" here will print "Belize" to the console.

### Customization

A `MacosSearchField` can be customized with almost all the `MacosTextField` properties, and additionally:

- An `emptyWidget` property, to display when there are no search results available.
- `resultHeight`, which controls the height of each result item.
- Via a `MacosSearchFieldTheme`, which controls the highlight and dropdown color.

Also, each `SearchResultItem` can also take a custom `child` widget and an explicit `onSelected` callback, for greater control. An example can be found in `fields_page.dart`.

Finally, this PR refactors a group of widgets that I found were used in various places to the new `MacosOverlayFilter` widget (not sure about the name!). It can be used to apply the blurry "frosted-glass" effect to surfaces, like the dropdowns, overlay menus, etc.. This caused the refactors in the pop-up, pull-down, and toolbar-overflow buttons.

### Future development

- Using the keyboard to navigate between search results.
- Add a toolbar version of this.

Closes #216.

## Pre-launch Checklist

- [👍 ] I have run `dartfmt` on all changed files 
- [👍  ] I have incremented the package version as appropriate and updated `CHANGELOG.md` with my changes 
- [ 👍 ] I have added/updated relevant documentation 
- [👍  ] I have run "optimize/organize imports" on all changed files
- [ 👍 ] I have addressed all analyzer warnings as best I could

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+## [1.0.1+1]
```
Comment (GroovinChip) [inline CHANGELOG.md]:
```suggestion
## [1.1.0]
```

File: lib/src/fields/search_field.dart [hunk 2]
```diff
@@ -50,23 +51,46 @@ class MacosSearchField extends StatelessWidget {
...
-    this.onEditingComplete,
-    this.onSubmitted,
     this.inputFormatters,
     this.enabled = true,
     this.onTap,
   }) : super(key: key);
 
-  /// Controls the text being edited.
+  /// List of results for the searchfield.
+  /// each suggestion should have a unique searchKey
```
Comment (GroovinChip) [inline lib/src/fields/search_field.dart]:
```suggestion
  /// 
  /// Each suggestion should have a unique searchKey.
```

File: lib/src/fields/search_field.dart [hunk 3]
```diff
@@ -50,23 +51,46 @@ class MacosSearchField extends StatelessWidget {
...
+  ///
+  /// ```dart
+  /// ['ABC', 'DEF', 'GHI', 'JKL']
+  ///   .map((e) => SearchResultItem(e, child: Text(e)))
+  ///   .toList(),
+  /// ```
+  final List<SearchResultItem>? results;
+
+  /// Callback when any suggestion is selected.
```
Comment (GroovinChip) [inline lib/src/fields/search_field.dart]:
```suggestion
  /// The action to perform when any suggestion is selected.
```

File: lib/src/fields/search_field.dart [hunk 4]
```diff
@@ -50,23 +51,46 @@ class MacosSearchField extends StatelessWidget {
...
+  final Function(SearchResultItem)? onResultSelected;
+
+  /// Specifies the number of results that will be displayed.
+  ///
+  /// It defaults to 5.
```
Comment (GroovinChip) [inline lib/src/fields/search_field.dart]:
```suggestion
  /// Defaults to `5`.
```

File: lib/src/fields/search_field.dart [hunk 5]
```diff
@@ -50,23 +51,46 @@ class MacosSearchField extends StatelessWidget {
...
+  final int maxResultsToShow;
+
+  /// Specifies height for each suggestion item in the list.
```
Comment (GroovinChip) [inline lib/src/fields/search_field.dart]:
```suggestion
  /// Specifies the height of each suggestion item in the list.
```

File: lib/src/fields/search_field.dart [hunk 6]
```diff
@@ -50,23 +51,46 @@ class MacosSearchField extends StatelessWidget {
...
   ///
-  /// If null, this widget will create its own [TextEditingController].
+  /// When not specified, the default value is `20.0`.
+  final double resultHeight;
+
+  /// Widget to show when the search returns no results.
+  ///
+  /// Defaults to [SizedBox.shrink]
```
Comment (GroovinChip) [inline lib/src/fields/search_field.dart]:
```suggestion
  /// Defaults to [SizedBox.shrink].
```

File: lib/src/theme/overlay_filter.dart [hunk 7]
```diff
@@ -0,0 +1,73 @@
+import 'dart:ui';
+import 'package:macos_ui/macos_ui.dart';
+import 'package:macos_ui/src/library.dart';
+
+/// Applies a blur filter to its child to create a macOS-style "frosted glass"
+/// effect.
```
Comment (GroovinChip) [inline lib/src/theme/overlay_filter.dart:7]:
```suggestion
/// {@template macosOverlayFilter}
/// Applies a blur filter to its child to create a macOS-style "frosted glass"
/// effect.
/// {@endtemplate}
```

File: lib/src/theme/overlay_filter.dart [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+class MacosOverlayFilter extends StatelessWidget {
+  /// Applies a blur filter to its child to create a macOS-style "frosted glass"
+  /// effect.
```
Comment (GroovinChip) [inline lib/src/theme/overlay_filter.dart]:
```suggestion
  /// {@macro macosOverlayFilter}
```

File: lib/src/theme/overlay_filter.dart [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    this.color,
+  }) : super(key: key);
+
+  final Widget child;
+
+  final BorderRadius borderRadius;
+
+  final double? height;
+
+  final Color? color;
```
Comment (GroovinChip) [inline lib/src/theme/overlay_filter.dart]: Would be nice for these to be documented

File: lib/src/theme/search_field_theme.dart [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  Widget wrap(BuildContext context, Widget child) {
+    return MacosSearchFieldTheme(data: data, child: child);
+  }
+
+  @override
+  bool updateShouldNotify(MacosSearchFieldTheme oldWidget) =>
+      data != oldWidget.data;
+}
+
+/// A style that overrides the default appearance of
```
Comment (GroovinChip) [inline lib/src/theme/search_field_theme.dart:50]:
```suggestion
/// {@template macosSearchFieldThemeData}
/// A style that overrides the default appearance of
```

File: lib/src/theme/search_field_theme.dart [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+/// [MacosSearchField]s when it is used with [MacosSearchFieldTheme] or with the
+/// overall [MacosTheme]'s [MacosThemeData.MacosSearchFieldTheme].
+///
+/// See also:
+///
+///  * [MacosSearchFieldTheme], the theme which is configured with this class.
+///  * [MacosThemeData.MacosSearchFieldTheme], which can be used to override the default
+///    style for [MacosSearchField]s below the overall [MacosTheme].
+class MacosSearchFieldThemeData with Diagnosticable {
```
Comment (GroovinChip) [inline lib/src/theme/search_field_theme.dart]:
```suggestion
///    style for [MacosSearchField]s below the overall [MacosTheme].
/// {@endtemplate}
class MacosSearchFieldThemeData with Diagnosticable {
```

File: lib/src/theme/search_field_theme.dart [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+  /// Creates a [MacosSearchFieldThemeData].
```
Comment (GroovinChip) [inline lib/src/theme/search_field_theme.dart]:
```suggestion
  /// {@macro macosSearchFieldThemeData}
```

File: lib/src/theme/search_field_theme.dart [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+  ///
+  /// Sets the color of a [SearchSuggestionItem]'s background when the mouse
+  /// hovers over it.
+  final Color? highlightColor;
+
+  /// The default background color for the [MacosSearchField] search results
+  /// overlay.
+  final Color? resultsBackgroundColor;
+
+  MacosSearchFieldThemeData copyWith({
```
Comment (GroovinChip) [inline lib/src/theme/search_field_theme.dart:78]:
```suggestion
  /// Copied one [MacosSearchFieldThemeData] to another.
  MacosSearchFieldThemeData copyWith({
```

File: pubspec.yaml [hunk 14]
```diff
@@ -1,6 +1,6 @@
 name: macos_ui
 description: Flutter widgets and themes implementing the current macOS design language.
-version: 1.0.1
+version: 1.0.1+1
```
Comment (GroovinChip) [inline pubspec.yaml]:
```suggestion
version: 1.1.0
```

**Review by GroovinChip — CHANGES_REQUESTED** [review]

File: lib/src/theme/overlay_filter.dart (hunk 9 above)
Comment (whiplashoo) [reply lib/src/theme/overlay_filter.dart]: In documenting these, I found that the height property can be omitted. The dimensions of the child widget are handled by it anyway; this is meant to be used as a decoration wrapper.

**Review by GroovinChip — APPROVED** [review]

**Pull request merged**